### PR TITLE
Basic auth community module

### PR DESCRIPTION
**Pull request by taba90**
This pull request add the basic auth community module. The module allows to authenticate users against an external service.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

**Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or not applicable.**

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [x] Make sure the first PR targets the `main` branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for community modules):
- [ ] There is a ticket in Jira describing the issue/improvement/feature (a notable exemptions is, changes not visible to end users)
- [ ] PR for bug fixes and small new features are presented as a single commit
- [ ] Commit message must be in the form "[GEOS-XYZW] Title of the Jira ticket" (export to XML in Jira generates the message in this exact form)
- [x] The pull request contains changes related to a single objective. If multiple focuses cannot be avoided, each one is in its own commit and has a separate ticket describing it.
- [ ] New unit tests have been added covering the changes
- [x] This PR passes all existing unit tests (test results will be reported by Continuous Integration after opening this PR)
- [x] This PR passes the [QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html) (QA checks results will be reported by Continuous Integration after opening this PR)
- [x] Commits changing the UI, existing user workflows, or adding new functionality, need to include documentation updates (screenshots, text)
- [ ] Commits changing the REST API, or any configuration object, should check if the REST API docs (Swagger YAML files and classic documentation) need to be updated.

### CODE REVIEW

**Review by aaime — CHANGES_REQUESTED** [review]
Please let's not call this "basic auth", it's confusing vs to what the module is actually doing. The security subsystem is complicated enough already, without adding extra confusion to existing terminology.

The basic authentication is one of the ways a username/password can be provided to a server, and it's already managed by a core functionality, in the form of an authentication filter. 

What the module contains is an authentication provider, which takes an already populated Authentication object (populated by an auth filter) and calls onto an external service to validate the username/password. This is not different from the LDAP authentication or the JDBC connection authentications. See for reference [the GeoSolutions training](https://docs.geoserver.geo-solutions.it/edu/en/security/authentication_providers.html#:~:text=Authentication%20Providers%20are%20pluggable%20components,in%20the%20standard%20GeoServer%20installation) and [the GeoServer own documentation](https://docs.geoserver.org/stable/en/user/security/auth/providers.html)

So please, can we call this "Web Service Authentication Provider" instead? And remove any reference to "basic" in the code and documentation?

File: doc/en/user/source/community/authbasic/index.rst [hunk 1]
```diff
@@ -0,0 +1,12 @@
+.. _community_authbasic:
+
+HTTP Based Authorization plug-in
+================================
+
+This plugin brings the possibility to authenticate users using an external authentication service.
```
[the remark this answers is not in the document]
Comment (taba90) [reply doc/en/user/source/community/authbasic/index.rst]: fixed

File: doc/en/user/source/community/authbasic/configuration.rst [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+HTTP Based Authorization configuration
+======================================
+
+
+Once the plug-in is installed reach the ``AuthenticationProvider`` from ``Authentication`` in the ``Security`` menu on the left and then click on ``add new`` in the ``AuthenticationProvider`` section. You will have the possibility to define a ``Web Service Authentication`` provider.
+
+.. figure:: images/newProvider.png
+
+Once clicked on ``Web Service Authentication`` you will have the possibility to enter the provider settings.
```
[the remark this answers is not in the document]
Comment (taba90) [reply doc/en/user/source/community/authbasic/configuration.rst]: fixed

File: doc/en/user/source/community/authbasic/configuration.rst [hunk 3]
```diff
@@ -0,0 +1,32 @@
+.. _authbasic_configuration:
+
+HTTP Based Authorization configuration
+======================================
+
+
+Once the plug-in is installed reach the ``AuthenticationProvider`` from ``Authentication`` in the ``Security`` menu on the left and then click on ``add new`` in the ``AuthenticationProvider`` section. You will have the possibility to define a ``Web Service Authentication`` provider.
```
[the remark this answers is not in the document]
Comment (taba90) [reply doc/en/user/source/community/authbasic/configuration.rst]: fixed

File: doc/en/user/source/community/authbasic/configuration.rst [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+Where:
+
+* ``Service URL`` is the URL of the external service meant to be used for authentication.
+
+* ``Timeout`` is the connection timeout.
+
+* ``Read Timeout`` is the timeout on waiting to read response data.
+
+* The ``Send credentials in X-HTTP-AUTHORIZATION Header`` checkbox is meant to be flagged if we want to send credential trough the authorization header. If unchecked (default) GeoServer will expect to find in the provided URL placeholders for username and password as ``{user}`` and ``{password}``.
```
[the remark this answers is not in the document]
Comment (taba90) [reply doc/en/user/source/community/authbasic/configuration.rst]: fixed

File: doc/en/user/source/community/authbasic/configuration.rst [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+* In the ``Authorization`` section the radio button allows to define whether to use a GeoServer ``RoleService`` to read roles or if roles are meant to be returned by the external authentication service.
+
+* In case ``Read Roles from Web Response`` is flagged you can specify a regex expression to extract the roles from the authentication service response.
```
[the remark this answers is not in the document]
Comment (taba90) [reply doc/en/user/source/community/authbasic/configuration.rst]: fixed

File: doc/en/user/source/community/authbasic/configuration.rst [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+Once the settings are saved the new ``AuthenticationProvider`` will be added to the list of the existing ones and you will be able to add it to the list of the providers' chain.
```
[the remark this answers is not in the document]
Comment (taba90) [reply doc/en/user/source/community/authbasic/configuration.rst]: fixed

File: doc/en/user/source/community/authbasic/configuration.rst [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+Once the plug-in is installed reach the ``AuthenticationProvider`` from ``Authentication`` in the ``Security`` menu on the left and then click on ``add new`` in the ``AuthenticationProvider`` section. You will have the possibility to define a ``Web Service Authentication`` provider.
+
+.. figure:: images/newProvider.png
+
+Once clicked on ``Web Service Authentication`` you will have the possibility to enter the provider settings.
+
+.. figure:: images/httpProviderConf.png
+
+Where:
```
[the remark this answers is not in the document]
Comment (taba90) [reply doc/en/user/source/community/authbasic/configuration.rst]: fixed

File: src/community/authbasic/src/main/java/org/geoserver/security/auth/web/SimpleWebAuthConfigException.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+ * This code is licensed under the GPL 2.0 license, available at the root
+ * application directory.
+ */
+package org.geoserver.security.auth.web;
+
+import org.geoserver.security.validation.SecurityConfigException;
+
+public class SimpleWebAuthConfigException extends SecurityConfigException {
+
+    public static final String INVALID_WEB_SERVICE_URL = "INVALID_WEB_SERVICE_URL";
```
[the remark this answers is not in the document]
Comment (taba90) [reply src/community/authbasic/src/main/java/org/geoserver/security/auth/web/SimpleWebAuthConfigException.java]: Those constants are mean to be used as key for the actual message defined in this properties file https://github.com/geoserver/geoserver/blob/3ab7bde62e4eb518e3247ea7600a133e7f0c4545/src/community/web-service-auth/src/main/resources/GeoServerException.properties

File: src/community/authbasic/src/main/java/org/geoserver/security/auth/web/SimpleWebServiceAuthenticationProvider.java [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+                                authentication.getPrincipal().toString())
+                        .replace(
+                                SimpleWebAuthenticationConfig.URL_PLACEHOLDER_PASSWORD,
+                                authentication.getCredentials().toString()));
+    }
+
+    private Map<String, String> getHeader(Authentication authentication) {
+        Map<String, String> headerMap = new HashMap<>();
+
+        String credentials = authentication.getPrincipal() + ":" + authentication.getCredentials();
```
[the remark this answers is not in the document]
Comment (taba90) [reply src/community/authbasic/src/main/java/org/geoserver/security/auth/web/SimpleWebServiceAuthenticationProvider.java]: fixed

File: src/community/authbasic/src/main/java/org/geoserver/security/auth/web/SimpleWebServiceAuthenticationProvider.java [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+                }
+            }
+        }
+        return new TreeSet<GrantedAuthority>(authorities);
+    }
+
+    // returns a URL with credentials substitued in place of place holders
+    private URL getAuthenticationURL(String connectionURL, Authentication authentication)
+            throws MalformedURLException {
+        return new URL(
```
[the remark this answers is not in the document]
Comment (taba90) [reply src/community/authbasic/src/main/java/org/geoserver/security/auth/web/SimpleWebServiceAuthenticationProvider.java]: fixed

File: src/community/authbasic/src/test/java/org/geoserver/security/auth/SimpleWebAuthProviderTest.java [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+        config.setRoleRegex("^.*?\"roles\"\\s*:\\s*\"([^\"]+)\".*$");
+
+        config.setAuthorizationOption(SimpleWebAuthenticationConfig.AUTHORIZATION_RADIO_OPTION_WEB);
+        config.setUseHeader(true);
+
+        getSecurityManager().saveAuthenticationProvider(config);
+
+        prepareFilterChain(pattern, testFilterName);
+        prepareAuthProviders(webAuthProviderUseHeader);
+        // SecurityContextHolder.getContext().setAuthentication(null);
```
[the remark this answers is not in the document]
Comment (taba90) [reply src/community/authbasic/src/test/java/org/geoserver/security/auth/SimpleWebAuthProviderTest.java]: fixed

File: src/community/authbasic/src/main/java/org/geoserver/security/auth/web/SimpleWebAuthenticationConfigValidator.java [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+
+        try {
+            new URL(connectionURL);
+        } catch (MalformedURLException e) {
+            LOGGER.log(Level.SEVERE, e.getMessage(), e);
+            throw new SimpleWebAuthConfigException(
+                    INVALID_WEB_SERVICE_URL, new Object[] {connectionURL});
+        }
+
+        if (!connectionURL.startsWith("https"))
```
[the remark this answers is not in the document]
Comment (taba90) [reply src/community/authbasic/src/main/java/org/geoserver/security/auth/web/SimpleWebAuthenticationConfigValidator.java]: fixed

File: src/community/authbasic/src/main/java/org/geoserver/security/auth/web/SimpleWebAuthenticationConfigValidator.java [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+            LOGGER.warning("Use of HTTPS is highly recommended");
+
+        // look for {user} and {password} if not set to use HEADER
+        if (!config.isUseHeader()) {
+            if (!(connectionURL
+                            .toLowerCase()
```
[the remark this answers is not in the document]
Comment (taba90) [reply src/community/authbasic/src/main/java/org/geoserver/security/auth/web/SimpleWebAuthenticationConfigValidator.java]: this one was hide by github, fixed but the code modification ended up to be in the refact commit.

**Comment by taba90** [comment]
thanks for your review @aaime your feedbacks have been applied

File: src/community/web-service-auth/src/main/java/org/geoserver/security/auth/web/WebServiceAuthenticationProvider.java [hunk 14]
```diff
@@ -0,0 +1,275 @@
...
+            roles.addAll(extractRoles(responseBody, config.getRoleRegex()));
+        }
+
+        // next extract user roles from configured service
+        // if no role service is selected, use the system default
+        try {
+            GeoServerRoleService roleService =
+                    getRoleService(authentication.getPrincipal().toString(), config);
+            roles.addAll(authorize(authentication.getPrincipal().toString(), roleService));
+            roles.addAll(checkAdminRoles(roles, roleService));
```
Comment (aaime) [inline src/community/web-service-auth/src/main/java/org/geoserver/security/auth/web/WebServiceAuthenticationProvider.java:265]: Not sure what's happening here, was it already there? If one of the roles is already recognized as admin role, why add it another time?

**Review by aaime — CHANGES_REQUESTED** [review]
Some more small feedback.

File: doc/en/user/source/community/web-service-auth/configuration.rst [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+
+The ``HTTP Based Authorization plug-in`` will try to authenticate the user on an configured external authentication service.
+The username and the password will be sent to the service in one of the following ways:
+
+* In a Header named ``X-HTTP-AUTHORIZATION``.
+* As a query parameters or as request path. For this use case the url needs to be configured by inseritng two placeholder, namely ``{user}`` and ``{password}`` , where the username and password are expected to be provided eg. ``https://my-auth-service?username={username}&password={password}``.
+
+The Authentication Provider will perform a ``GET`` request, sending credential Base64 encoded. If the response status returned by the external service is different from ``200`` the user will not be authenticated.
+
+In case the external authentication service is returning the authenticated user's roles in the response body, it is possible to define a regex expression to extract them, allowing  for their usage for authorization. There is no limitation to a specific content type.
```
[the remark this answers is not in the document]
Comment (taba90) [reply doc/en/user/source/community/web-service-auth/configuration.rst]: fixed

File: src/community/web-service-auth/src/main/java/org/geoserver/security/auth/web/WebAuthenticationConfig.java [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+        return allowHTTPConnection;
+    }
+
+    public void setAllowHTTPConnection(boolean allowHTTPConnection) {
+        this.allowHTTPConnection = allowHTTPConnection;
+    }
+
+    @Override
+    public String toString() {
+        return "SimpleWebAuthenticationConfig [connectionURL="
```
[the remark this answers is not in the document]
Comment (taba90) [reply src/community/web-service-auth/src/main/java/org/geoserver/security/auth/web/WebAuthenticationConfig.java]: fixed

File: src/community/web-service-auth/src/main/java/org/geoserver/security/auth/web/WebAuthenticationConfigValidator.java [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+
+    Set<String> availableRoles;
+
+    public WebAuthenticationConfigValidator(GeoServerSecurityManager securityManager) {
+        super(securityManager);
+        try {
+            availableRoles = securityManager.listRoleServices();
+        } catch (IOException e) {
+            LOGGER.log(
+                    Level.SEVERE, "Error in SimpleWebAuthenticationConfigValidator constructor", e);
```
[the remark this answers is not in the document]
Comment (taba90) [reply src/community/web-service-auth/src/main/java/org/geoserver/security/auth/web/WebAuthenticationConfigValidator.java]: fixed

File: src/community/web-service-auth/src/main/java/org/geoserver/security/auth/web/WebServiceAuthenticationProvider.java (hunk 14 above)
Comment (taba90) [reply src/community/web-service-auth/src/main/java/org/geoserver/security/auth/web/WebServiceAuthenticationProvider.java:253]: The admin authority is assigned by the the ROLE_ADMIN. The additional check is used to check whether any of the role already returned by the query by username is equal to the admin role known from the roleService if it is then add the ROLE_ADMIN constants. Similar to what happens [here](https://github.com/geoserver/geoserver/blob/de184b2f9ed9a844b5291dd163e772132a07689b/src/main/src/main/java/org/geoserver/security/impl/RoleCalculator.java#L111)

**Review by aaime — APPROVED** [review]

**Comment by taba90** [comment]
jira ticket https://osgeo-org.atlassian.net/browse/GEOS-10010

**Comment by petersmythe** [comment]
It appears as though this new community module is not being built.  Instructions at https://docs.geoserver.org/latest/en/user/community/web-service-auth/install.html result in 404.

**Comment by aaime** [comment]
Community modules are officially unsupported so don't hold your breath about it :-D
That said, the packaging instructions are there and the module is actually built, from the logs:

```
[INFO] Building zip: /var/lib/jenkins/workspace/geoserver-master/src/community/target/release/geoserver-2.20-SNAPSHOT-cov-json-plugin.zip
[INFO] Building zip: /var/lib/jenkins/workspace/geoserver-master/src/community/target/release/geoserver-2.20-SNAPSHOT-mongodb-schemaless-plugin.zip
[INFO] Building zip: /var/lib/jenkins/workspace/geoserver-master/src/community/target/release/geoserver-2.20-SNAPSHOT-web-service-auth.zip
[INFO] Building zip: /var/lib/jenkins/workspace/geoserver-master/src/community/target/release/geoserver-2.20-SNAPSHOT-gwc-mbtiles-plugin.zip
```

Ah! It has not been named following conventions, so the nightly build script is not copying it over to the nightly build distribution:
``cp community/target/release/geoserver-*-plugin.zip $dist/community-latest``
but the module is not named "geoserver-*-plugin.zip" like the others. @taba90

**Pull request merged**